internal/core/textlayout/sharedparley.rs

```diff
@@ -0,0 +1,138 @@
+};
+use i_slint_common::sharedfontique;
+
+pub const DEFAULT_FONT_SIZE: LogicalLength = LogicalLength::new(12.);
+
+fn font_context() -> parley::FontContext {
+    parley::FontContext {
+        collection: sharedfontique::COLLECTION.inner.clone(),
+        source_cache: sharedfontique::COLLECTION.source_cache.clone(),
+    }
+}
+
+static LAYOUT_CONTEXT: std::sync::LazyLock<parley::LayoutContext<Brush>> =
+    std::sync::LazyLock::new(|| Default::default());
+
+#[derive(Debug, Default, PartialEq, Clone, Copy)]
+pub struct Brush {
+    pub is_selected: bool,
+    pub stroke: Option<TextStrokeStyle>,
+}
+
+pub struct LayoutOptions {
+    pub max_width: Option<LogicalLength>,
+    pub max_height: Option<LogicalLength>,
+    pub horizontal_align: TextHorizontalAlignment,
+    pub vertical_align: TextVerticalAlignment,
+    pub stroke: Option<TextStrokeStyle>,
+    pub selection: Option<std::ops::Range<usize>>,
+    pub font_request: Option<FontRequest>,
+    pub text_wrap: TextWrap,
+    pub text_overflow: TextOverflow,
+}
+
+impl Default for LayoutOptions {
+    fn default() -> Self {
+        Self {
+            max_width: None,
+            max_height: None,
+            horizontal_align: TextHorizontalAlignment::Left,
+            vertical_align: TextVerticalAlignment::Top,
+            stroke: None,
+            selection: None,
+            font_request: None,
+            text_wrap: TextWrap::WordWrap,
+            text_overflow: TextOverflow::Clip,
+        }
+    }
+}
+
+pub fn layout(text: &str, scale_factor: f32, options: LayoutOptions) -> Layout {
+    let mut font_context = font_context();
+    let mut layout_context = LAYOUT_CONTEXT.clone();
+
+    let mut builder = layout_context.ranged_builder(&mut font_context, text, scale_factor, true);
+    if let Some(ref font_request) = options.font_request {
+        if let Some(family) = &font_request.family {
+            builder.push_default(parley::StyleProperty::FontStack(
+                parley::style::FontStack::Single(parley::style::FontFamily::Named(
+                    family.as_str().into(),
+                )),
+            ));
+        }
+        if let Some(weight) = font_request.weight {
+            builder.push_default(parley::StyleProperty::FontWeight(
+                parley::style::FontWeight::new(weight as f32),
+            ));
+        }
+        if let Some(letter_spacing) = font_request.letter_spacing {
+            builder.push_default(parley::StyleProperty::LetterSpacing(letter_spacing.get()));
+        }
+        builder.push_default(parley::StyleProperty::FontStyle(if font_request.italic {
+            parley::style::FontStyle::Italic
+        } else {
+            parley::style::FontStyle::Normal
+        }));
+    }
+    let pixel_size = options
+        .font_request
+        .and_then(|font_request| font_request.pixel_size)
+        .unwrap_or(DEFAULT_FONT_SIZE);
+    builder.push_default(parley::StyleProperty::FontSize(pixel_size.get()));
```

Comment:
The incoming pixel size is a LogicalLength. Shouldn't it be multiplied with the `scale_factor` before passing it to parley?

Reply:
Where is a scale factor passed to parley?

AFAICS the font size in parley is in what we'd call physical pixels, and parley seems to use it as the factor to scale from the font metrics - so all good with that one AFAICS. Is there another scale factor that I missed?